docs/architecture.md

```diff
@@ -263,3 +270,46 @@ one of or many:
 
 * internal [`client`](#client) library
 * third-party client library
+
+# Implementation <a name="implementation"/>
+
+For convenience and clarity, the code should be written in a functional programming style. It's mandatory to avoid the OOP style which almost always makes things much more complicated than should be.
+
+## Public API <a name="implementation_public_api"/>
+
+Any application or library always have a public API and internal/private part. For convenience keep public functions and structures in the root package, so the API will be easy to find. Additionally, if the `component`:
+
+* is providing reach public API with additional structures. - Is mandatory to distinct the public structures and functions from internal implementation which should be kept in nested package(s).
+* is just a simple tool with a compact implementation that is not specifying many structures. - Private implementation can be kept in the same file, just behind the public API or even whole tool can be delivered as one public function if the implementation is simple enough.
+
+DO NOT keep many public functions along with internal implementation in the same file or package, because it is messing up the public API, which makes code harder to analyze and navigate.
```

Comment:
```suggestion
DO NOT keep multiple public functions along with internal implementations in the same file or package, because it is messes up the public API, which makes code harder to analyze and navigate.
```